Adding MoleculerJS #1124

AndreMaz opened the pull request:
This PR is an attempt to add [Moleculer](https://moleculer.services/) to the list of back-end frameworks https://github.com/microsoft/WebTemplateStudio/issues/1124

What was done/changed:
- Default Template containing:
  - 2 Moleculer microservices
    - HTTP gateway service. *Note*: 404 handler for nonexistent endpoints is present by default
    - `Pages` service containing the application logic
  - `Readme.md` + links pointing to Moleculer docs for more info
  - Fully documented `moleculer.config.js` for backend tweaking
- Composition Templates for the following requests:
  - `GET /grid`
  - `GET /list`
  - `POST /list`
  - `DELETE /list/:_id`
  - `GET /masterdetail`
- Client updates
  - Had to update some files in `src/client/src` to add the `.svg` icon to the Wizard

What's not done:
- Integration with Azure Cosmos


Some prints (copy&paste from #1124)
Wizard: 
![image](https://user-images.githubusercontent.com/9802754/64132980-50d98a80-cdcb-11e9-9c74-fa8d0e55f2c6.png)
Moleculer+Vue
![FM2eNnx8T4](https://user-images.githubusercontent.com/9802754/64133011-7e263880-cdcb-11e9-9fff-9ea4ff764ad0.gif)

AndreMaz committed:
[moleculer] Add MoleculerJS to backend

AndreMaz committed:
[moleculer] Use auto-alias in template

AndreMaz committed:
[moleculer] Adding composition templates

AndreMaz committed:
[moleculer] add moleculer logo to client

AndreMaz committed:
[moleculer] cleaning things

AndreMaz committed:
Merge remote-tracking branch 'upstream/dev' into dev

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/microsoft/WebTemplateStudio?pullRequest=1173) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/WebTemplateStudio?pullRequest=1173) before we can accept your contribution.<br/><br/>:x: AndreMaz [sign now](https://cla.opensource.microsoft.com/microsoft/WebTemplateStudio?pullRequest=1173)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/microsoft/WebTemplateStudio?pullRequest=1173) it.</sub>

AndreMaz committed:
[moleculer] fix missing error import

smmatte commented on src/client/src/mockData/mockVsCodeApi.ts:
```diff
@@ -186,6 +186,23 @@ const mockVsCodeApi = () => ({
+                    author: "Various",
+                    order: "1",
+                    icon:
+                      "https://raw.githubusercontent.com/moleculerjs/branding/master/logo/logo-nocircle.svg",
```
we should be consistent with the other frameworks and leave this icon empty

AndreMaz replied:
Nice catch. I've removed

AndreMaz committed:
[moleculer] remove icon entry

AndreMaz commented:
Hi everyone

I found something weird. Apparently the sequence by which the pages are selected (Step 3 - 
Add Pages) affect the template.

Context:
I have 3 composition templates that require `sampleData`
![Untitled](https://user-images.githubusercontent.com/9802754/64478257-ce234780-d19d-11e9-8f1b-cd08813b5620.png)

One with:
```js
//{[{
const sampleData = require("../data/sampleData");
const { MoleculerError } = require("moleculer").Errors;
//}]}
```
and other two with: 
```js
//{[{
const sampleData = require("../data/sampleData");
//}]}
```

Now check how the merging algorithm works when I select
- List -> Grid -> Master (works properly)
![list-grid-master](https://user-images.githubusercontent.com/9802754/64478285-28bca380-d19e-11e9-8a46-450ba5d92bbc.gif)

I get 2 `require` 
```js
const sampleData = require("../data/sampleData");
const { MoleculerError } = require("moleculer").Errors;
```
___
- Grid -> List -> Master (wrong)
![grid-list-master](https://user-images.githubusercontent.com/9802754/64478308-7cc78800-d19e-11e9-8309-7747dd97f863.gif)

I get 3 `require` 
```js
const sampleData = require("../data/sampleData");
const { MoleculerError } = require("moleculer").Errors;

const sampleData = require("../data/sampleData"); // Repeated
```
___
- Master -> List -> Grid (wrong)
![master-list-grid](https://user-images.githubusercontent.com/9802754/64478316-9d8fdd80-d19e-11e9-9359-6fcd11981f87.gif)

I get 3 `require` 
```js
const sampleData = require("../data/sampleData");
const { MoleculerError } = require("moleculer").Errors;

const sampleData = require("../data/sampleData"); // Repeated
```

Is this something that I did or this is a bug? Can somebody help out on this one?

sibille commented:
@AndreMaz, this happens due to how the merge postactions work. The code between //{[{ and //}]} is added if the block to insert is not there already (to avoid duplication). The block is analyzed as a whole. I'll have a look at your templates to see how to reorganize to make the additions work correctly.

AndreMaz commented:
Hey @sibille 

> The code between //{[{ and //}]} is added if the block to insert is not there already (to avoid duplication). The block is analyzed as a whole.

Ok. Since the blocks are analyzed as a whole and I have a template like this for `List Page`:
```js
//{[{
const sampleData = require("../data/sampleData");
const { MoleculerError } = require("moleculer").Errors;
//}]}
```
and like this for `Master Page` and `Grid Page`:
```js
//{[{
const sampleData = require("../data/sampleData");
//}]}
```
then the merging algorithm wouldn't be able to detect that `sampleData` has already been required and it would add it again.

So changing  `List Page`  template to something like this:
```js
//{[{
const sampleData = require("../data/sampleData");
//{[{
//}]}
const { MoleculerError } = require("moleculer").Errors;
//}]}
```
would solve the issue. This time the `sampleData` block is equal in all templates so it should only be included once.

However, the end result is exactly the same (wrong) as `sampleData` is included twice.

sibille commented:
Although you splitted the additions into two blocks it's still managed as a single block because there are no contextual lines between the two block. What you can do to solve your problem is extract the common part (the require sample data) to a common composition template. I did a quick test in this branch, could you have a look at it and see if that works as expected for you? https://github.com/sibille/WebTemplateStudio/tree/test_molecular/

AndreMaz committed:
[moleculer] fix repeated sampleData include

Followed @sibille recommendation:
https://github.com/sibille/WebTemplateStudio/tree/test_molecular/

AndreMaz commented:
Thanks @sibille :smile: now everything works properly

jcoc611-microsoft commented on src/client/src/assets/moleculer.svg:
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="utf-8"?>
+<svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
+	 viewBox="0 0 512 512" style="enable-background:new 0 0 512 512;" xml:space="preserve">
```
For accessibility, could you add the following attributes:

* `alt` brief description of this image
* `aria-hidden="true"` so screen readers ignore this image

And the following element in the body of the `<svg>`:

```xml
<svg>
    <title>Moleculer Logo</title>
    ...
</svg>
```

AndreMaz replied:
Done https://github.com/microsoft/WebTemplateStudio/pull/1173/commits/5e962fa4222cdd7036a40ef014e5714bb1e8c246

jcoc611-microsoft commented on src/client/src/mockData/mockVsCodeApi.ts:
```diff
@@ -186,6 +186,21 @@ const mockVsCodeApi = () => ({
+                    languages: ["Any"],
+                    tags: {
+                      version: "0.13.10",
+                      preview: false
```
If you'd like to deploy a "beta" (preview) version of Moleculer first, you can set this one to `true`. Otherwise we will need to spend more time to make sure is ready for a prod release.

AndreMaz replied:
Done https://github.com/microsoft/WebTemplateStudio/pull/1173/commits/5e962fa4222cdd7036a40ef014e5714bb1e8c246

AndreMaz replied:
@jcoc611-microsoft let me know if you need anything else 😉

AndreMaz committed:
[moleculer]changes requested by @jcoc611-microsoft

jcoc611-microsoft commented on templates/_catalog/backendframeworks.json:
```diff
@@ -26,5 +26,19 @@
+    "languages": ["Any"],
+    "tags": {
+      "version": "0.13.10",
+      "preview": false
```
Sorry, could you also change this one to true? Thank you!

AndreMaz replied:
Done https://github.com/microsoft/WebTemplateStudio/pull/1173/commits/c09d6307b959b4ae9573d06b7bec4f2df9514a40

AndreMaz committed:
[moleculer] preview: true

dgomezc commented:
Hi @AndreMaz, the molecular-repl package contains a dependency on lodash package that has two vulnerabilities.

| High | Prototype Pollution |
| --- | --- |
| Package | lodash |
| Patched in | >=4.17.11 |
| Dependency of | moleculer-repl [dev] |
| Path | moleculer-repl > vorpal > inquirer > lodash |
| More info | https://nodesecurity.io/advisories/782 |

| High | Prototype Pollution |
| --- | --- |
| Package | lodash |
| Patched in | >=4.17.12 |
| Dependency | of moleculer-repl [dev] |
| Path | moleculer-repl > vorpal > inquirer > lodash |
| More info | https://nodesecurity.io/advisories/1065 |

Can you fix this before merging?

AndreMaz committed:
[moleculer] drop unnecessary dev dependency

AndreMaz commented:
@dgomezc done. 
The vulnerability was caused by a unnecessary dev dependency that I've completely removed.
![image](https://user-images.githubusercontent.com/9802754/64976651-c2283b80-d8a9-11e9-945a-f6cd578488b6.png)

AndreMaz committed:
Revert previous commit.

Things changed:
- Solved all vulnerabilities in dev dependency. So I've brought it back
- Renamed handlers. Made them clearer

AndreMaz commented:
I've solved the vulnerabilities pointed out by @dgomezc in https://github.com/microsoft/WebTemplateStudio/pull/1173#issuecomment-531817021

In https://github.com/microsoft/WebTemplateStudio/pull/1173/commits/b174ab1ee1686e1aa25329e5ebaf0844c02a0ee8 I've completely removed the dev dependency (`moleculer-repl`) from the template. This was quick and dirty fix and I'm not proud of it :unamused:

In meantime, I've managed to solve those vulnerabilities. So in https://github.com/microsoft/WebTemplateStudio/pull/1173/commits/9c53a37f0f96653ab475a99c7a87fbb616403e27 I've bought `moleculer-repl` back. 

`moleculer-repl` is an (optional) interactive developer console.  While optional it's extremelly useful as 
we can use it to interact with the microservices (e.g., check their status, load or destroy them, call services, etc.) without the HTTP gateway.

Some prints:
**List services**
![image](https://user-images.githubusercontent.com/9802754/65070930-78ad1e80-d985-11e9-960e-d9702a8a3057.png)

**Call service action directly from the terminal**
![image](https://user-images.githubusercontent.com/9802754/65071005-a1351880-d985-11e9-928a-1d309459d4cd.png)


**List nodes data**
![image](https://user-images.githubusercontent.com/9802754/65071062-bf027d80-d985-11e9-8786-22a903806f3a.png)


@dgomezc @sibille @smmatte @markAtMicrosoft @jcoc611-microsoft can you review this PR again?

AndreMaz committed:
[moleculer} Prettier --fix

smmatte commented:
@AndreMaz thank you for your work, I think that the branch is ready to be merged in dev.

AndreMaz commented:
Thanks everybody for helping me throughout the whole process :smile: 

I'm already thinking about next PR. Can somebody take a look at https://github.com/microsoft/WebTemplateStudio/issues/1124#issuecomment-531571687 ?
